feat: BROS-347: Vector

nick-skriabin opened the pull request:
Introducing Vector annotations

This pull request introduces significant improvements and new features for vector region support in the editor, especially focusing on bezier curve editing, programmatic manipulation, and selection management. It also adds new UI components and updates schema/tag support for polyline and vector shapes.

**Vector Region and Bezier Curve Enhancements:**

- Added a comprehensive `KonvaVector` component README detailing new programmatic point creation and transformation APIs, including bezier curve support, bounding box calculations, and a singleton PointCreationManager for consistent point handling.
- Introduced `ControlPoints.tsx` and `ConnectionLines.tsx` components to visually handle bezier control points and connection lines for vector editing. [[1]](diffhunk://#diff-6eab87e8dbeb9507fdff41de5c5fe2221370f9f1612df4409d9c6e8351cfc154R1-R107) [[2]](diffhunk://#diff-9e711347900eb4114d56354ecdf4613492736e59c7c3210ffa718c016b330d05R1-R72)
- Implemented a global `VectorSelectionTracker` singleton to manage point selection, active instances, and transformation state across all vector instances in the editor.

**Schema and Tag Support:**

- Updated `tags.json` to add "Polyline" and "PolylineLabels" for expanded vector annotation support.

**Integration and Region Handling:**

- Updated region splitting and rendering logic in `ImageView.jsx` to support new vector region types, ensuring they are handled distinctly from other region types. [[1]](diffhunk://#diff-9e437cca91d1dff57dcf03b57c1a642bff8c6b5513277e2c28cc04e7094d582aR50) [[2]](diffhunk://#diff-9e437cca91d1dff57dcf03b57c1a642bff8c6b5513277e2c28cc04e7094d582aR69) [[3]](diffhunk://#diff-9e437cca91d1dff57dcf03b57c1a642bff8c6b5513277e2c28cc04e7094d582aL77-R81) [[4]](diffhunk://#diff-9e437cca91d1dff57dcf03b57c1a642bff8c6b5513277e2c28cc04e7094d582aL1437-R1449)
- Enhanced region detection and watcher logic to recognize and correctly process `vectorregion` types in overlays and connectors. [[1]](diffhunk://#diff-90c91b60046b0b57fdfc7d81075411f883a078c3df01aa68e3e048693c69a694R126) [[2]](diffhunk://#diff-19130085f2d864c56833f4e42111d0fee1c0a61f7cd0d4bd3a8c6e4cf42a9eabR40-R41)

**SVG Allowlist Adjustment:**

- Modified the SVG element allowlist, replacing `polyline` with `vector` to align with updated vector region handling.

These changes collectively provide a robust foundation for advanced vector annotation workflows, including bezier editing, multi-point selection, and programmatic manipulation.

---

**Vector Region and Bezier Curve Enhancements**
- Added detailed documentation for `KonvaVector`, including APIs for programmatic point creation, bezier support, and shape transformations.
- Introduced `ControlPoints.tsx` for bezier control point visualization and `ConnectionLines.tsx` for skeleton connections. [[1]](diffhunk://#diff-6eab87e8dbeb9507fdff41de5c5fe2221370f9f1612df4409d9c6e8351cfc154R1-R107) [[2]](diffhunk://#diff-9e711347900eb4114d56354ecdf4613492736e59c7c3210ffa718c016b330d05R1-R72)
- Implemented `VectorSelectionTracker` singleton for managing global selection and transformation state across vector instances.

**Schema and Tag Support**
- Added "Polyline" and "PolylineLabels" to `tags.json` for improved schema support.

**Integration and Region Handling**
- Updated region splitting, rendering, and detection logic to support new vector region types and ensure correct integration in overlays and connectors. [[1]](diffhunk://#diff-9e437cca91d1dff57dcf03b57c1a642bff8c6b5513277e2c28cc04e7094d582aR50) [[2]](diffhunk://#diff-9e437cca91d1dff57dcf03b57c1a642bff8c6b5513277e2c28cc04e7094d582aR69) [[3]](diffhunk://#diff-9e437cca91d1dff57dcf03b57c1a642bff8c6b5513277e2c28cc04e7094d582aL77-R81) [[4]](diffhunk://#diff-9e437cca91d1dff57dcf03b57c1a642bff8c6b5513277e2c28cc04e7094d582aL1437-R1449) [[5]](diffhunk://#diff-90c91b60046b0b57fdfc7d81075411f883a078c3df01aa68e3e048693c69a694R126) [[6]](diffhunk://#diff-19130085f2d864c56833f4e42111d0fee1c0a61f7cd0d4bd3a8c6e4cf42a9eabR40-R41)

**SVG Allowlist Adjustment**
- Changed SVG allowlist to use `vector` instead of `polyline`.

<img width="1850" height="1634" alt="image" src="https://github.com/user-attachments/assets/a3c0ef4b-f4fc-49e8-ae93-4f76af19c013" />

<!--

This description MUST be filled out for a PR to receive a review. Its primary purposes are:

 - to enable your reviewer to review your code easily, and
 - to convince your reviewer that your code works as intended.

Some pointers to think about when filling out your PR description:
 - Reason for change: Description of problem and solution
 - Screenshots: All visible changes should include screenshots.
 - Rollout strategy: How will this code be rolled out? Feature flags / env var / other
 - Testing: Description of how this is being verified
 - Risks: Are there any known risks associated with this change, eg to security or performance?
 - Reviewer notes: Any info to help reviewers approve the PR
 - General notes: Any info to help onlookers understand the code, or callouts to significant portions.

You may use AI tools such as Copilot Actions to assist with writing your PR description (see https://docs.github.com/en/copilot/using-github-copilot/using-github-copilot-for-pull-requests/creating-a-pull-request-summary-with-github-copilot); however, an AI summary isn't enough by itself. You'll need to provide your reviewer with strong evidence that your code works as intended, which requires actually running the code and showing that it works.

-->

nick-skriabin committed:
Polyline placeholder

nick-skriabin committed:
Typescript

nick-skriabin committed:
Add KonvaVector

nick-skriabin committed:
Semi-working

nick-skriabin committed:
Various behavior and transformation fixes

nick-skriabin committed:
Behavior fixes

nick-skriabin committed:
Shape events (wip)

nick-skriabin committed:
Migrate from Shape to Path to unblock multi-layer interactionso

- Fixed Path scaling
- Fixed alt key interactions
- Fixed rendering issues when extra segments were added to closed shape

nick-skriabin committed:
Support configuration options, behavior fixes

nick-skriabin committed:
Bounds

nick-skriabin committed:
Transformer constraints

nick-skriabin committed:
Remove redundant docs

nick-skriabin committed:
Skeleton fixes, fix closing behavior when there are just two points

nick-skriabin commented:
/fmt
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/17256927561)

nick-skriabin committed:
Apply pre-commit linters

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17256927561

nick-skriabin committed:
Skeleton, close path, point addition

nick-skriabin committed:
Merge branch 'develop' into fb-bros-347/vector

nick-skriabin committed:
Change points addition from alt to shift

nick-skriabin committed:
Migrato off of isPathClosed to ref-based closing

nick-skriabin committed:
Change double click to cmd-click to select all points

nick-skriabin committed:
Path breaking

nick-skriabin committed:
Restore couple of files

nick-skriabin committed:
Proper path closing

nick-skriabin committed:
Adding points

nick-skriabin committed:
Drawing vector region

nick-skriabin committed:
Point manager, immediate drawing

nick-skriabin committed:
Continuous drawing, proper path closure behavior

nick-skriabin committed:
Selection

nick-skriabin committed:
Selection

nick-skriabin committed:
Programmatic transform

nick-skriabin commented:
/fmt
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/17285422800)

nick-skriabin committed:
Transformer pt.2

nick-skriabin committed:
Relations

nick-skriabin commented:
/fmt
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/17286296783)

nick-skriabin committed:
Biome format

nick-skriabin committed:
Fix lint

nick-skriabin committed:
Bezier conversion algo

nick-skriabin committed:
Konva 18

nick-skriabin committed:
Proper polyline support

nick-skriabin committed:
Point styles

nick-skriabin committed:
Biome

yyassi-heartex commented on web/libs/editor/src/components/KonvaVector/KonvaVector.tsx:
```diff
@@ -0,0 +1,1606 @@
+
+  // Determine if drawing should be disabled based on current interaction context
+  const isDrawingDisabled = () => {
+    // Disable all interactions when disabled prop is true
+    if (disabled) {
+      return true;
+    }
+
+    // Disable drawing when Shift is held (for Shift+click functionality)
+    if (isShiftKeyHeld) {
+      return true;
+    }
+
+    // Disable drawing when multiple points are selected
+    if (selectedPoints.size > 1) {
+      return true;
+    }
```
```suggestion
    // Disable all interactions when disabled prop is true
    if (disabled || isShiftKeyHeld || selectedPoints.size > 1) {
      return true;
    }
```

yyassi-heartex commented on web/libs/editor/src/components/KonvaVector/KonvaVector.tsx:
```diff
@@ -0,0 +1,1606 @@
+      const incomplete = minPoints !== undefined && allPoints.length < minPoints;
+
+      const shapeData = {
+        type: (allowClose ? "polygon" : "polyline") as "polygon" | "polyline",
```
please create an enum or a type then update usage everywhere we use "polygon" or "polyline". this will help reduce the chance of mistakes relating to spelling plus help document the code a little bit better

yyassi-heartex commented on web/libs/editor/src/components/KonvaVector/components/ConnectionLines.tsx:
```diff
@@ -0,0 +1,72 @@
+          <Line
+            key={`connection-${branch.id}`}
+            points={[startPoint.x, startPoint.y, firstBranchPoint.x, firstBranchPoint.y]}
+            stroke="#3b82f6"
```
nitpick: I see this stroke colour being hardcoded and used in other spots. lets create a constant and reuse it from there so we only need to update it in one spot

yyassi-heartex commented on web/libs/editor/src/components/KonvaVector/components/ControlPoints.tsx:
```diff
@@ -0,0 +1,107 @@
+
+        return (
+          <React.Fragment
+            key={`control-group-${index}-${point.x.toFixed(2)}-${point.y.toFixed(2)}-${point.controlPoint1?.x.toFixed(2) || "null"}-${point.controlPoint1?.y.toFixed(2) || "null"}-${point.controlPoint2?.x.toFixed(2) || "null"}-${point.controlPoint2?.y.toFixed(2) || "null"}`}
```
maybe we should memoize this string so it doesn't get rebuilt on every re-render

nick-skriabin replied:
it will still change every time point's moved. shouldn't be an issue

yyassi-heartex commented on web/libs/editor/src/components/KonvaVector/components/ControlPoints.tsx:
```diff
@@ -0,0 +1,107 @@
+              <Rect
+                x={point.controlPoint1.x}
+                y={point.controlPoint1.y}
+                width={scaledRadius * 1.6}
+                height={scaledRadius * 1.6}
+                offsetX={scaledRadius * 0.8}
+                offsetY={scaledRadius * 0.8}
```
lets set the 1.6 and 0.8 as constants since they seem to be meaningful ratios - will help the code be self documenting

yyassi-heartex commented on web/libs/editor/src/components/KonvaVector/components/GhostLine.tsx:
```diff
@@ -0,0 +1,289 @@
+
+    const firstPoint = initialPoints[0];
+    const lastPoint = initialPoints[initialPoints.length - 1];
+    const closeRadius = 15 / (transform.zoom * fitScale);
```
what is 15 representing here?

nick-skriabin committed:
Optional chaining for incomplete in outliner tree (could actually be undefined)

nick-skriabin committed:
Revert to develop state

bmartel commented on web/libs/editor/src/examples/image_polylines/annotations/0.json:
Rename all polylines references to vector.

nick-skriabin committed:
Constants, history fix

nick-skriabin committed:
Enums

nick-skriabin committed:
More constants

nick-skriabin committed:
Even more constants

nick-skriabin committed:
Stroke color

nick-skriabin committed:
Simplify expression

nick-skriabin committed:
Formatting

nick-skriabin committed:
Formatting

nick-skriabin committed:
Formatting

nick-skriabin committed:
Selection and path closing behavior

nick-skriabin committed:
Remove polyline example

nick-skriabin committed:
Vector docs

nick-skriabin committed:
More styling options

nick-skriabin committed:
Formatting

nick-skriabin committed:
Formatting

nick-skriabin committed:
Update output format

nick-skriabin committed:
Docs + formatting

nick-skriabin committed:
Add structure docs

nick-skriabin committed:
Add bezier docs

nick-skriabin committed:
Merge branch 'develop' into fb-bros-347/vector

nick-skriabin committed:
Jsx :(

Docs generator doesn't support tsx

robot-ci-heartex committed:
ci: Build tag docs

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17310142847

hlomzik commented on web/libs/editor/src/tags/control/VectorLabels.jsx:
```diff
@@ -0,0 +1,87 @@
+  SelectedModelMixin.props({ _child: "LabelModel" }),
+);
+
+const VectorLabelsModel = types.compose("VectorLabelsModel", Composition);
```
small inconsistency: MST supports and we usually use simple composition with name along with no extra steps:
```
const VectorLabelsModel = types.compose("VectorLabelsModel", ControlBase, LabelsModel, ...)
```

hlomzik commented on web/libs/editor/src/components/KonvaVector/README.md:
This readme mentions "new something" couple times like it was just an addition to another document. Do you have it or that's just some glitches?

hlomzik replied:
Also this readme doesn't cover coords system used in Vector. And I see that you added `internalToImageX()` and similar methods. Do you store coords in image coords? Does it have any benefits comparing to our internal system?

nick-skriabin replied:
it's cursor messing around. it was updating the document along the way. will fix, ty

the coordinates are in image coordinates, yes. it was just much easier to manipulate all the Konva stuff directly in pixels instead of constantly converting to %. 

considering that KonvaVector is almost a third party self-contained library that we integrate with LSO, i think it's ok that it uses pixels and image coordinates directly

nick-skriabin committed:
Cleanup

nick-skriabin committed:
Fix readme

nick-skriabin committed:
Relative coords

nick-skriabin committed:
Document relative and image coords

nick-skriabin committed:
Update the doc

robot-ci-heartex committed:
ci: Build tag docs

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17325247386

nick-skriabin committed:
Conversion

bmartel commented on web/libs/editor/src/tools/Vector.js:
```diff
@@ -0,0 +1,242 @@
+      createRegionOptions() {
+        return Super.createRegionOptions({
+          shape: [],
+          // shape: [],
```
```suggestion
```

nick-skriabin committed:
Relative points covnersions

bmartel commented on web/libs/editor/src/components/KonvaVector/components/GhostLine.tsx:
```diff
@@ -0,0 +1,290 @@
+  const closingTarget = getClosingTarget();
+  const shouldRender = cursorPosition && !isPathClosed && (!drawingDisabled || closingTarget);
+
+  // Debug logging removed
```
```suggestion
```

hlomzik commented on web/libs/editor/src/regions/VectorRegion.jsx:
```diff
@@ -50,6 +50,9 @@ const Model = types
     readonly: types.optional(types.boolean, false),
+
+    // Internal flag to detect if we converted data back from relative points
+    converted: false,
```
I'd move it to volatile

hlomzik replied:
ah, ok, I see you are using it and it actually might be helpful.
nevermind then!

robot-ci-heartex committed:
ci: Build tag docs

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17325898114

hlomzik commented on web/libs/editor/src/tags/visual/View.jsx:
```diff
@@ -99,6 +99,7 @@ const Model = types
       "rectanglelabels",
       "ellipselabels",
       "polygonlabels",
+      "vectorlabels",
```
`"vector"` is missing in places like this. add it please

nick-skriabin replied:
done

bmartel commented on docs/source/includes/tags/vectorlabels.md:
```diff
@@ -0,0 +1,51 @@
+| original_height | <code>number</code> | height of the original image (px) |
+| image_rotation | <code>number</code> | rotation degree of the image (deg) |
+| value | <code>Object</code> |  |
+| value.shape | <code>Array.&lt;Object&gt;</code> | array of point objects with coordinates, bezier curve information, and point relationships |
```
```suggestion
| value.vertices | <code>Array.&lt;Object&gt;</code> | array of point objects with coordinates, bezier curve information, and point relationships |
```

I thought we were going to rename this `shape` value everywhere? I was also thinking that instead of `vectorpoints`, `vertices` could be a more fitting term to use which would not collide with existing data keys and also be more descriptive. It could also be argued that `vectorpoints` is more consistent with the way we name other things like `vectorlabels`, so that still would make sense. I would say for sure we don't want to ship with `shape`, as that is a generic term that doesn't actually communicate what data it holds. Most often this could be quickly confused with expecting the value type to be a string name of a shape (like `rectangle`, `polygon` etc.).

hlomzik replied:
@nick-skriabin your recent commits covers not all the occurencies. Like this one the comment attached to.

nick-skriabin replied:
yeah, just did it today's morning. `vertices` it is

nick-skriabin committed:
Rename shape to vectorPoints

hlomzik reviewed (approved):
I can't comprehend the whole logic of `KonvaVector`. But all integration points look good, making all possible issues at least isolated and incapsulated within this new tag. I mean points such as:
- Adding tags, regions and results
- Coords conversion including undo/redo
- All related tags are aknowledged of new tags and regions (except for some places missing `Vector` tag, added a comment about this)

robot-ci-heartex committed:
ci: Build tag docs

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17326206984

nick-skriabin committed:
Update web/libs/editor/src/components/KonvaVector/components/GhostLine.tsx

Co-authored-by: bmartel <brandonmartel@gmail.com>

nick-skriabin committed:
vectorPoints → vertices

robot-ci-heartex committed:
ci: Build tag docs

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17327426804

nick-skriabin committed:
Formatting
